crate/operator/main.py

```diff
@@ -382,191 +374,59 @@ async def cluster_create(
-                ],
-                new_data_replicas=[
-                    WebhookScaleNodePayload(
-                        name=item["name"], replicas=item["replicas"]
-                    )
-                    for item in body.spec["nodes"]["data"]
-                ],
-                old_master_replicas=old["spec"]["nodes"]
-                .get("master", {})
-                .get("replicas"),
-                new_master_replicas=body.spec["nodes"]
-                .get("master", {})
-                .get("replicas"),
-            )
-            try:
-                await with_timeout(
-                    scale_cluster(
-                        apps,
-                        core,
-                        namespace,
-                        name,
-                        do_scale_data,
-                        do_scale_master,
-                        get_total_nodes_count(old["spec"]["nodes"]),
-                        spec,
-                        scale_master_diff_item,
-                        (
-                            kopf.Diff(scale_data_diff_items)
-                            if scale_data_diff_items
-                            else None
-                        ),
-                        logger,
-                    ),
-                    config.SCALING_TIMEOUT,
-                    (
-                        f"Failed to scale cluster {namespace}/{name} after "
-                        f"{config.SCALING_TIMEOUT} seconds."
-                    ),
-                )
-            except Exception:
-                logger.exception("Failed to scale cluster")
-                await webhook_client.send_scale_notification(
-                    WebhookStatus.FAILURE,
-                    namespace,
-                    name,
-                    webhook_scale_payload,
-                    logger,
-                )
-                raise
-            else:
-                logger.info("Cluster scaled")
-                await webhook_client.send_scale_notification(
-                    WebhookStatus.SUCCESS,
-                    namespace,
-                    name,
-                    webhook_scale_payload,
-                    logger,
-                )
+    if do_upgrade:
+        if new_cycle:
+            context.state_machine.add(State.UPGRADE)
+        kopf.register(fn=UpgradeSubHandler(namespace, name, context)(), id="upgrade")
+
+    if do_restart:
+        if new_cycle:
+            context.state_machine.add(State.RESTART)
+        kopf.register(
+            fn=RestartSubHandler(namespace, name, context)(),
+            id="restart",
+            timeout=config.ROLLING_RESTART_TIMEOUT,
+        )
+
+    if do_scale:
+        if new_cycle:
+            context.state_machine.add(State.SCALE)
+        kopf.register(
+            fn=ScaleSubHandler(namespace, name, context)(),
+            id="scale",
+            timeout=config.SCALING_TIMEOUT,
+        )
+
+    patch.status["updateContext"] = context.serialize()
```

Comment:
This whole function looks very clean and well structured! Good work!

Reply:
:blush: Thank you!